File: server/VideoStream.cpp
#include "VideoStream.h"
#include <stdexcept>
#include <iostream>

VideoStream::VideoStream(std::string fName)
    : videoFile(), fileName(std::move(fName)), frameNbr(0)
{
    if (!fileName.empty()) {
        openFile(fileName);
    }
}

void VideoStream::openFile(const std::string& filename)
{
    if (videoFile.is_open()) videoFile.close();
    videoFile.open(filename, std::ios::binary);
    if (!videoFile || videoFile.fail() || !videoFile.is_open()) {
        throw std::runtime_error("Failed to open video file: " + filename);
    }
    fileName = filename;
    frameNbr = 0;
}

// Read next JPEG frame from a MJPEG-style file by scanning for 0xFFD8 ... 0xFFD9
int VideoStream::getNextFrame(uint8_t* frameBuf, int bufSize)
{
    if (!frameBuf || bufSize <= 0) return -1;
    if (!videoFile.is_open()) return -1;

    int prev = -1;
    int cur = -1;

    // Find JPEG start marker 0xFF 0xD8
    while ((cur = videoFile.get()) != EOF) {
        if (prev == 0xFF && cur == 0xD8) {
            // start of frame found; write the two start bytes
            int written = 0;
            if (bufSize < 2) return -1;
            frameBuf[written++] = 0xFF;
            frameBuf[written++] = 0xD8;

            int prevByte = 0xD8; // last written byte value
            int byteRead;
            // Read until end marker 0xFF 0xD9 (inclusive)
            while ((byteRead = videoFile.get()) != EOF) {
                if (written >= bufSize) {
                    // buffer too small: skip until end marker, then return error
                    int p = prevByte;
                    int q;
                    while ((q = videoFile.get()) != EOF) {
                        if (p == 0xFF && q == 0xD9) break;
                        p = q;
                    }
                    return -1;
                }
                frameBuf[written++] = static_cast<uint8_t>(byteRead);
                if (prevByte == 0xFF && byteRead == 0xD9) {
                    ++frameNbr;
                    return written; // complete frame size
                }
                prevByte = byteRead;
            }
            return -1; // EOF before end marker
        }
        prev = cur;
    }

    // No more frames
    return -1;
}           

File: server/video_stream.py
from __future__ import annotations

from pathlib import Path
from typing import Optional, BinaryIO


JPEG_START = b"\xFF\xD8"
JPEG_END = b"\xFF\xD9"


class VideoStream:
    def __init__(self, path: Path) -> None:
        self._path = path
        self._file: Optional[BinaryIO] = None
        self._open_file(path)

    def _open_file(self, path: Path) -> None:
        if self._file:
            self._file.close()
        self._file = path.open("rb")

    def get_next_frame(self) -> Optional[bytes]:
        if not self._file:
            return None

        prev = b""
        while True:
            cur = self._file.read(1)
            if not cur:
                return None
            if prev + cur == JPEG_START:
                frame = bytearray(JPEG_START)
                prev_byte = JPEG_START[1:2]
                while True:
                    byte = self._file.read(1)
                    if not byte:
                        return None
                    frame += byte
                    if prev_byte + byte == JPEG_END:
                        return bytes(frame)
                    prev_byte = byte
            prev = cur
